### PR TITLE
Use base class name for reification of associations

### PR DESCRIPTION
**Pull request by andre1810**
Use base class name for reification instead of class name to reify polymorphic models properly.

Refers to #594 but won´t fix all issues. If you have more `has_one` associations to an polymorphic model you won´t get the correct associated models because sub type reference is missing in query.

### CODE REVIEW

File: lib/paper_trail/reifiers/belongs_to.rb [hunk 1]
```diff
@@ -39,7 +39,7 @@ def load_record(assoc, id, options, version)
         # @api private
         def load_version(assoc, id, transaction_id, version_at)
           assoc.klass.paper_trail.version_class.
-            where("item_type = ?", assoc.klass.name).
+            where("item_type = ?", assoc.klass.base_class.name).
```
Comment (jaredbeck) [inline lib/paper_trail/reifiers/belongs_to.rb]: Nice. I was hoping AR provided something like this `base_class`, but I couldn't find it.

**Comment by jaredbeck** [comment]
This looks good André. Please add:

1. at least one test, perhaps based on https://gist.github.com/andre1810/3cc1cbfaa8344a8ad8a840fbb9f91890, but let's use existing test models if we can. How about `Person`, `Animal`, `Cat`, and `Dog`? Person has many pets?
2. a changelog entry, under 9.0.0 -> Fixed (once this is in `master` we can discuss a backport if needed)

> .. [this] won´t fix all issues. If you have more has_one associations to an polymorphic model you won´t get the correct associated models because sub type reference is missing in query.

Do we have a bug report demonstrating that? Is there a gist I can run? Should we open a new issue for that or continue discussing it in https://github.com/airblade/paper_trail/issues/594?

**Comment by andre1810** [comment]
@jaredbeck I added a test to verify the changes regarding `has_many` associations and another one to show the still open issue #594 with `has_one` associations which is currently skipped. If you don´t want to include the issue I can remove it but I think it will help solving the issue later. Maybe we should keep #594 open because it´s not fixed at all.

I also added a changelog entry. Backward compatibility would be nice as I currently use 8.1.1.

File: spec/models/person_spec.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+require "spec_helper"
+
+RSpec.describe Person, type: :model, versioning: true do
+  it "baseline test setup" do
+    expect(Person.new).to be_versioned
+  end
+
+  describe "#cars and bicycles" do
+    it "can be reified", skip: "Known Issue #594" do
```
Comment (jaredbeck) [inline spec/models/person_spec.rb]: This is great, thanks for including it.

File: spec/models/pet_spec.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    last_version = person.reload.versions.last.reify(has_many: true)
+    expect(last_version.pets.length).to(eq(2))
+    expect(last_version.animals.length).to(eq(2))
+    expect(last_version.animals.map { |a| a.class.name }).to(eq(%w[Dog Cat]))
+    expect(last_version.pets.map { |p| p.animal.class.name }).to(eq(%w[Dog Cat]))
+    expect(last_version.animals.first.name).to(eq("Beethoven"))
+    expect(last_version.dogs.first.name).to(eq("Beethoven"))
+    expect(last_version.animals.second.name).to(eq("Sylvester"))
+    expect(last_version.cats.first.name).to(eq("Sylvester"))
```
Comment (jaredbeck) [inline spec/models/pet_spec.rb]: This test reads well. :+1:

File: spec/dummy_app/db/migrate/20110208155312_set_up_test_tables.rb [hunk 4]
```diff
@@ -36,17 +36,18 @@ def up
...
+      t.integer :owner_id
+      t.timestamps null: false, limit: 6
     end
 
     create_table :skippers, force: true do |t|
       t.string     :name
-      t.datetime   :another_timestamp
-      t.timestamps null: true
+      t.datetime   :another_timestamp, limit: 6
+      t.timestamps null: true, limit: 6
```
Comment (jaredbeck) [inline spec/dummy_app/db/migrate/20110208155312_set_up_test_tables.rb]: Ugh. MySQL's lack of Fractional Seconds Precision (FSP) has been a big headache on this project. I'm OK with this change, but can you remind me what datatype this produces in MySQL? Is it a `DATETIME(6)`? Does that mean six decimal points of FSP?

I wonder if, after this change, we can use less `timecop` in our tests .. but that's a story for another time. :smile:

**Comment by jaredbeck** [comment]
Can you organize this into two commits, the first adding fractional seconds precision and the second doing everything else? If it's a pain, don't worry about it, I'll just squash it into one commit when I merge.

**Pull request closed without merging**